itests/clients/net_client.go

```diff
@@ -0,0 +1,222 @@
+	defer c.closedLock.Unlock()
+	c.t.Logf("Closing connection to %s node at %q (%t)", c.impl.String(), c.s.RemoteAddr().String(), c.closedFlag)
+	if c.closedFlag {
+		return
+	}
+	_ = c.s.Close()
+	c.closedFlag = true
+}
+
+func (c *NetClient) reconnect() {
+	c.t.Logf("Reconnecting to %q", c.s.RemoteAddr().String())
+	conn, err := net.Dial("tcp", c.s.RemoteAddr().String())
+	require.NoError(c.t, err, "failed to dial TCP to %s node", c.impl.String())
+
+	s, err := c.n.NewSession(c.ctx, conn, c.c)
+	require.NoError(c.t, err, "failed to re-establish the session to %s node", c.impl.String())
+	c.s = s
+
+	c.SendHandshake()
+}
+
+func (c *NetClient) closing() bool {
+	c.closingLock.Lock()
+	defer c.closingLock.Unlock()
+	return c.closingFlag
+}
+
+type protocol struct {
+	dropLock sync.Mutex
+	drop     map[proto.PeerMessageID]struct{}
+}
+
+func newProtocol(drop []proto.PeerMessageID) *protocol {
+	m := make(map[proto.PeerMessageID]struct{})
+	for _, id := range drop {
+		m[id] = struct{}{}
+	}
+	return &protocol{drop: m}
+}
+
+func (p *protocol) EmptyHandshake() networking.Handshake {
+	return &proto.Handshake{}
+}
+
+func (p *protocol) EmptyHeader() networking.Header {
+	return &proto.Header{}
+}
+
+func (p *protocol) Ping() ([]byte, error) {
+	msg := &proto.GetPeersMessage{}
+	return msg.MarshalBinary()
+}
+
+func (p *protocol) IsAcceptableHandshake(h networking.Handshake) bool {
+	hs, ok := h.(*proto.Handshake)
+	if !ok {
+		return false
+	}
+	// Reject nodes with incorrect network bytes, unsupported protocol versions,
+	// or a zero nonce (indicating a self-connection).
+	if hs.AppName != appName || hs.Version.Cmp(proto.ProtocolVersion()) < 0 || hs.NodeNonce == 0 {
+		return false
+	}
+	return true
+}
+
+func (p *protocol) IsAcceptableMessage(h networking.Header) bool {
+	hdr, ok := h.(*proto.Header)
+	if !ok {
+		return false
+	}
+	p.dropLock.Lock()
+	defer p.dropLock.Unlock()
+	_, ok = p.drop[hdr.ContentID]
+	return !ok
+}
+
+type handler struct {
+	peers  []proto.PeerInfo
+	t      testing.TB
+	client *NetClient
+}
+
+func newHandler(t testing.TB, peers []proto.PeerInfo) *handler {
+	return &handler{t: t, peers: peers}
+}
+
+func (h *handler) OnReceive(s *networking.Session, data []byte) {
+	msg, err := proto.UnmarshalMessage(data)
+	if err != nil { // Fail test on unmarshal error.
+		h.t.Logf("Failed to unmarshal message from bytes: %q", base64.StdEncoding.EncodeToString(data))
+		h.t.FailNow()
+		return
+	}
+	switch msg.(type) { // Only reply with peers on GetPeersMessage.
+	case *proto.GetPeersMessage:
+		h.t.Logf("Received GetPeersMessage from %q", s.RemoteAddr())
+		rpl := &proto.PeersMessage{Peers: h.peers}
+		bts, mErr := rpl.MarshalBinary()
+		if mErr != nil { // Fail test on marshal error.
+			h.t.Logf("Failed to marshal peers message: %v", mErr)
+			h.t.FailNow()
+			return
+		}
+		if _, wErr := s.Write(bts); wErr != nil {
+			h.t.Logf("Failed to send peers message: %v", wErr)
+			h.t.FailNow()
+			return
+		}
+	default:
+	}
```

Comment:
Not now. Later I'll add the messages expectations to the protocol. So, it will be possible to expect certain messages and fail on not receiving them.

Reply:
For now `NetClient` resembles the old one, so it mainly used to send messages without noticing replies.